here-naksha-app-service/src/test/java/com/here/naksha/app/service/ValDryRunTest.java

```diff
@@ -0,0 +1,153 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+package com.here.naksha.app.service;
+
+import static com.here.naksha.app.common.CommonApiTestSetup.*;
+import static com.here.naksha.app.common.ResponseAssertions.assertThat;
+import static com.here.naksha.app.common.TestUtil.*;
+import static org.junit.jupiter.api.Assertions.*;
+
+import com.here.naksha.app.common.ApiTest;
+import com.here.naksha.app.common.NakshaTestWebClient;
+import com.here.naksha.app.service.models.FeatureCollectionRequest;
+import com.here.naksha.lib.core.models.geojson.implementation.XyzFeature;
+import com.here.naksha.lib.core.models.geojson.implementation.XyzFeatureCollection;
+import com.here.naksha.lib.core.models.geojson.implementation.XyzReference;
+import com.here.naksha.lib.core.models.naksha.Space;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.http.HttpResponse;
+import java.util.List;
+import java.util.UUID;
+import org.jetbrains.annotations.NotNull;
+import org.json.JSONException;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.skyscreamer.jsonassert.JSONAssert;
+import org.skyscreamer.jsonassert.JSONCompareMode;
+import org.skyscreamer.jsonassert.comparator.ArraySizeComparator;
+
+public class ValDryRunTest extends ApiTest {
+
+  private static final NakshaTestWebClient nakshaClient = new NakshaTestWebClient();
+
+  private static final String SPACE_ID = "local-space-4-val-dry-run";
+
+  @BeforeAll
+  static void setup() throws URISyntaxException, IOException, InterruptedException {
+    createHandler(nakshaClient, "ValDryRun/setup");
+    createSpace(nakshaClient, "ValDryRun/setup");
+  }
+
+  private void additionalCustomAssertions_tc3000(final @NotNull String reqBody, final @NotNull String resBody)
+      throws JSONException {
+    final FeatureCollectionRequest collectionRequest = parseJson(reqBody, FeatureCollectionRequest.class);
+    final XyzFeatureCollection collectionResponse = parseJson(resBody, XyzFeatureCollection.class);
+    final List<String> updatedIds = collectionResponse.getUpdated();
+    final List<XyzFeature> features = collectionResponse.getFeatures();
+    final List<XyzFeature> violations = collectionResponse.getViolations();
+    JSONAssert.assertEquals(
+        "{updated:[" + collectionRequest.getFeatures().size() + "]}",
+        resBody,
+        new ArraySizeComparator(JSONCompareMode.LENIENT));
+    assertEquals(
+        updatedIds.size(), features.size(), "Mismatch between updated and features list size in the response");
```

Comment:
Why do we check size twice (L66-69; L70-71)?

Reply:
the first one checks against feature size in original request. other one checks for array size within response itself. so, there is difference. but will see if it can be optimized